lib/dockerregistry/filestore.go

```diff
@@ -0,0 +1,408 @@
+	if err != nil {
+		return nil, fmt.Errorf("error building GCS client: %v", err)
+	}
+
+	prefix := strings.TrimPrefix(u.Path, "/")
+	if prefix != "" && !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+
+	bucket := u.Host
+
+	s := &gcsSyncFilestore{
+		filestore: filestore,
+		client:    client,
+		bucket:    bucket,
+		prefix:    prefix,
+	}
+	return s, nil
+}
+
+// OpenReader opens an io.ReadCloser for the specified file
+func (s *gcsSyncFilestore) OpenReader(ctx context.Context, name string) (io.ReadCloser, error) {
+	absolutePath := s.prefix + name
+	return s.client.Bucket(s.bucket).Object(absolutePath).NewReader(ctx)
+}
+
+// UploadFile uploads a local file to the specified destination
+func (s *gcsSyncFilestore) UploadFile(ctx context.Context, dest string, localFile string) error {
+	absolutePath := s.prefix + dest
+
+	gcsUrl := "gs://" + s.bucket + "/" + absolutePath
+
+	in, err := os.Open(localFile)
+	if err != nil {
+		return fmt.Errorf("error opening %q: %v", localFile, err)
+	}
+	defer in.Close()
+
+	klog.Infof("uploading to %s", gcsUrl)
+
+	out := s.client.Bucket(s.bucket).Object(absolutePath).NewWriter(ctx)
+
+	if _, err := io.Copy(out, in); err != nil {
+		out.Close() // best effort
+		// TODO: Try to delete the possibly partially written file?
+		return fmt.Errorf("error uploading to %q: %v", gcsUrl, err)
+	}
+
+	if err := out.Close(); err != nil {
+		return fmt.Errorf("error uploading to %q: %v", gcsUrl, err)
+	}
+
+	return nil
+}
+
+// ListFiles returns all the all artifacts in the filestore, recursively.
+func (s *gcsSyncFilestore) ListFiles(ctx context.Context) (map[string]*syncFileInfo, error) {
+	files := make(map[string]*syncFileInfo)
+
+	q := &storage.Query{Prefix: s.prefix}
+	klog.Infof("listing files in bucket %s with prefix %s", s.bucket, s.prefix)
+	it := s.client.Bucket(s.bucket).Objects(ctx, q)
+	for {
+		obj, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("error listing objects in %q: %v", s.filestore.Base, err)
+		}
+		name := obj.Name
+		if !strings.HasPrefix(name, s.prefix) {
+			return nil, fmt.Errorf("found object %q without prefix %q", name, s.prefix)
+		}
+
+		file := &syncFileInfo{}
+		file.AbsolutePath = "gs://" + s.bucket + "/" + obj.Name
+		file.RelativePath = strings.TrimPrefix(name, s.prefix)
+		if obj.MD5 == nil {
+			return nil, fmt.Errorf("MD5 not set on file %q", file.AbsolutePath)
+		} else {
+			file.MD5 = hex.EncodeToString(obj.MD5)
+		}
+		file.Size = obj.Size
+		file.filestore = s
+
+		files[file.RelativePath] = file
+	}
+
+	return files, nil
+}
+
+// syncFileOp defines a synchronization operation
+type syncFileOp interface {
+	Run(ctx context.Context) error
+}
+
+// computeNeededOperations determines the list of files that need to be copied
+func (p *FilestorePromoter) computeNeededOperations(source, dest map[string]*syncFileInfo, destFilestore syncFilestore) ([]syncFileOp, error) {
```

Comment:
It will be much nicer if we can reuse (or at least emulate) the set-difference logic we use for figuring out what needs to be promoted (copied), as in https://github.com/kubernetes-sigs/k8s-container-image-promoter/blob/f2a8a6e4557beada0a6cebd45a9c8d2b8433d951/lib/dockerregistry/set.go. For example we could just serialize the metadata for the file into a single string key, which would be the value in a map (the keys would be the filenames). Then the logic to figure out what files need to be copied is just a set minus operation, as in https://github.com/kubernetes-sigs/k8s-container-image-promoter/blob/edca50185d79a18857b21e2ae8bdb4d346b0887e/lib/dockerregistry/inventory.go#L1067.

Reply:
Given the new upcoming changes for how the promoter works, this argument for reusing set-difference logic no longer holds true. I think the set-difference stuff will be used later in a different tool that performs audits/backups of entire registries/repositories.